sycl/doc/design/DeviceIf.md

```diff
@@ -338,4 +338,1011 @@ in that compilation phase.
+`N2`, etc. telling the _Condition Expression_ for the _Conditional Actions_ in
+this _Add On Image_.  The property's size (which is always divisible by 4)
+tells the number of `uint32` integers in this property.
+
+```
+[SYCL/add on images]
+"1": N1, N2, N3, ...
+"2": N1, N2, N3, ...
+```
+
+Each _Add On Image_ contains a property set named "SYCL/add on image" which
+contains just one property.  The name of the property is the unique identifier
+for this _Add On Image_ (e.g. "1").  The value of the property has type
+`PI_PROPERTY_TYPE_UINT32`.  The value is `1` if the _Add On Image_ contains the
+real function definitions, and it is `0` if the _Add On Image_ contains the
+stub definitions.
+
+```
+[SYCL/add on image]
+"1": 1
+```
+
+### Changes to the DPC++ runtime
+
+The DPC++ runtime requires changes to identify and dynamically link the _Add On
+Images_ when necessary.
+
+When the application submits a kernel to a device, the runtime must already
+find the device images from shared libraries that define exported device
+functions as described in [Device Code Dynamic Linking][6].  This algorithm is
+extended to look also for _Add On Images_.
+
+If the _Main Image_ contains the "SYCL/add on images" property set, the runtime
+does the following for each property in that set:
+
+* The name of the property is the ID for a pair of _Add On Images_.  The
+  runtime searches the other device images looking at their "SYCL/add on image"
+  property sets in order to find the _Add On Images_ with the same ID.  There
+  should be two such images: a primary image and a stub image.
+
+* The value of the property is the _Condition Expression_ gating the
+  _Conditional Actions_ in this _Add On Image_.  The runtime evaluates this
+  expression for the device to which the kernel is being submitted.
+
+* The runtime builds a set of _Add On Images_ to link against.  If the
+  expression is true, the runtime adds the primary _Add On Image_ to this set.
+  If the expression is false, the runtime adds the stub _Add On Image_ to this
+  set.
+
+* The selected _Add On Image_ may also contain a "SYCL/add on images" property
+  set naming yet more _Add On Image_ pairs.  If so, the runtime adds these
+  properties to the iteration set, causing their _Add On Images_ to be found
+  also.
+
+Once this completes, the runtime computes the union of the
+"SYCL/device requirements" property sets from the _Main Image_, all the _Add On
+Images_, and any images from shared libraries defining exported functions.  The
+runtime then uses this combined set of device requirements to check if the
+kernel is compatible with the device as described in the
+[Optional Device Features][7] design.
+
+If the kernel is compatible, the runtime creates a `pi_program` by compiling
+and linking all these device images together.
+
+[6]: <./SharedLibraries.md>
+[7]: <./OptionalDeviceFeatures.md#changes-to-the-dpc-runtime>
+
+
+## Alternate design for non-AOT SPIR-V targets
```

Comment:
Personally, I'm in favor of the main proposed design for non-AOT SPIR-V targets, because similar idea is likely to be used by virtual functions extension (#10540) and therefore we will have more code which is handled in a similar manner and less things to support.

Raising a bar for underlying backends is not a huge concern in here, because we are talking about an extension and not a core functionality. However, if we think that this would be promoted to core SYCL at some point, then it is maybe better to avoid requiring a SPIR-V extension for this.